org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryManager.java

```diff
@@ -1057,21 +1044,16 @@ MavenProject getMavenProject(final MavenProjectFacade facade, final IProgressMon
       return mavenProject;
     }
     Map<MavenProjectFacade, MavenProject> mavenProjects = getContextProjects();
-    mavenProject = mavenProjects.get(facade);
-    if(mavenProject == null) {
-      try {
-        mavenProject = mavenProjectCache.get(facade,
-            () -> readProjectWithDependencies(facade.getPom(), facade.getResolverConfiguration(), monitor));
-      } catch(ExecutionException ex) {
-        Throwable cause = ex.getCause();
-        if(cause instanceof CoreException) {
-          throw (CoreException) cause;
-        }
-        throw new RuntimeException(cause); // this really should never happen
+    try {
+      return mavenProjects.computeIfAbsent(facade, fac -> mavenProjectCache.computeIfAbsent(fac,
+          f -> readProjectWithDependencies(f.getPom(), f.getResolverConfiguration(), monitor)));
+    } catch(IllegalStateException ex) { // thrown by method called in lambda to carry CoreException
```

Comment:
IllegalStateException is really bad choice I think... WHy not catch runtme exceptions at all and look if the cause if a CoreException, if not create a new one with "internal error" message?

Reply:
That's right. Done.